src/Plutarch/Extra/DScript.hs

```diff
@@ -0,0 +1,196 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Plutarch.Extra.DScript (
+    DScript (..),
+    checkedCompileD,
+    mustCompileD,
+    mustFinalEvalDScript,
+    finalEvalDScript,
+    mustEvalScript,
+    mustEvalD,
+) where
+
+import Control.DeepSeq (NFData)
+import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text as Text
+import GHC.Generics (Generic)
+import Plutarch (
+    Config (Config, tracingMode),
+    TracingMode (DetTracing, NoTracing),
+    compile,
+ )
+import Plutarch.Evaluate (EvalError, evalScript)
+import Plutarch.Extra.Compile (mustCompile)
+import Plutarch.Prelude (S, Term, Type)
+import PlutusLedgerApi.V1 (ExBudget, Script)
+import UntypedPlutusCore.Evaluation.Machine.Cek (
+    CekUserError (CekEvaluationFailure, CekOutOfExError),
+    ErrorWithCause (ErrorWithCause),
+    EvaluationError (InternalEvaluationError, UserEvaluationError),
+ )
+
+-- | A 'Script' with a debug fallback that has tracing turned on.
+data DScript = DScript {script :: Script, debugScript :: Script}
```

Comment:
I guess the `D` refers to `debug` here? It'd be a bit clearer if it had that in its name, along with all its functions.

Reply:
I'd prefer `DebugScript`

Reply:
I thought about that, wanted to avoid having it sound like a `DebugScript` is only for debugging. I'm not sure about this. Any name that better explains what's going on is too long, like `ScriptWithDebugFallback`.

Reply:
I'm digging through a thesaurus, and I'm coming up with 

- `ProvisionalScripts` (i.e., a script that is executed normally with the provisio that tracing is turned on in certain circumstances). I like this because "provisional" isn't a super over-loaded term, so most people won't come away with misunderstanding just from a glance
- "ContingencyScripts" 
- "ConditionalScripts" -- but this one uses the (very) overloaded "Conditional".

Reply:
How about `DebuggableScript`?